[Create-block] Add `--variant` flag to allow creation of different block type variants

ryanwelcher opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This PR introduces a new CLI flag to the `create-block` tool to indicate if the generated block should be dynamic or static. The default state is to create a static block and by passing `--variant dynamic` a dynamic block would be created.

## Why?
There is currently no way to scaffold a dynamic block using `create-block` without using a custom template. The intent here is to allow any template, including the built-in ones, to support creating either type of block.

## How?
I have added the new flag, `--variant`, and am currently using a combination of Mustache template conditionals and custom-named templates to load the correct files.

## Testing Instructions
1. Run `node index.js my-test --variant dynamic` from the root of the create-block package
2. Ensure that there is a `template.php` file added to the src directory and that no `save.js` file was added or is referenced in `index.js`
3. Run `node index.js my-test ` from the root of the create-block package and confirm that `save.js` is in-place and references correctly in `index.js`

ryanwelcher committed:
Add npmDevDependencies as a template variable. Update messaging to show what packages are being installed.

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Remove extra code block.

ryanwelcher committed:
Add --is-dynamic flag and some conditional logic as a first pass at managing both block type options.

ryanwelcher committed:
Update templates and tutorial packages to use new templates.

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge branch 'trunk' into try/adding-is-dynamic-flag

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge branch 'trunk' of github.com:ryanwelcher/gutenberg into trunk

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge branch 'trunk' into try/adding-is-dynamic-flag

ryanwelcher committed:
Revert code to check the slug of a file.

ryanwelcher committed:
Check for empty template files before writing them into the filesystem.

ryanwelcher committed:
Remove files with block type in the name.

ryanwelcher committed:
Add the {{isDynamic}} placeholders to the templates bundled with the package to manage output based on the --is-dynamic flag.

ryanwelcher commented:
Based on a conversation with @gziolo during contributor day at WCEU, this PR now uses the {{isDynamic}} template placeholder to conditionally out parts of the templates.

It also now checks to see if a file is empty before writing it to the filesystem. This allows template authors to handle the use-case of files that should only be included depending on the type of block by allowing them to wrap the content in  the conditional placeholders:
```
{{#isDynamic}}
Only render if --is-dynamic is passed
{{/isDynamic}}

{{^isDynamic}}
Only render if --is-dynamic is NOT passed
{{/isDynamic}}
```

ryanwelcher committed:
Add readme and changelog entries.

ryanwelcher committed:
Render the output files will whitespace maintained.

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

gziolo commented on packages/create-block/lib/output.js:
```diff
@@ -19,7 +19,12 @@ const writeOutputTemplate = async ( inputFile, outputFile, view ) => {
 	await makeDir( dirname( outputFilePath ) );
-	writeFile( outputFilePath, render( inputFile, view ) );
+	// If the rendered template is empty, don't write it. This is how we can conditionally add template files.
+	const renderedFile = render( inputFile, view ).trim();
+	if ( renderedFile.length ) {
+		// We need to re-render so that the whitespace at the end of files are maintained
+		writeFile( outputFilePath, render( inputFile, view ) );
+	}
```
This would work better:
```suggestion
	const renderedFile = render( inputFile, view );
	if ( renderedFile.trim().length ) {
		// We need to re-render so that the whitespace at the end of files are maintained
		writeFile( outputFilePath, renderedFile );
	}
```

ryanwelcher replied:
I did something similar here and when the files were output ( specifically the .scss ) they were triggering linting errors due to missing trailing spaces.

gziolo replied:
The trick is that I propose to run trimming only when evaluating the condition in the `if` statement. The content of the rendered file should remain unchanged when saving to the file.

ryanwelcher replied:
ah! I see! That _IS_ a much better way. I'll update the PR.

gziolo commented on packages/create-block/lib/templates/plugin/$slug.php.mustache:
```diff
@@ -37,7 +37,35 @@
+ */
+function {{namespaceSnakeCase}}_{{slugSnakeCase}}_render_callback( $atts, $content, $block) {
+	ob_start();
+	require plugin_dir_path( __FILE__ ) . 'build/template.php';
```
Does the PHP file gets copied by default or do you need to pass the flag to the `npm run build` command?

gziolo replied:
Maybe, we should inline the template as the first step and focus on better integration with render templates in WordPress core next?

ryanwelcher replied:
The build will require the flag to copy the files over. I really like the idea of having the template defined elsewhere - perhaps the block.json. ACF is doing something similar  - https://github.com/AdvancedCustomFields/acf/issues/654

In the meantime, I'm going to add the flag to the generated scripts based on the existence of the `--is-dynamic` flag.

gziolo replied:
That works for now, too 👍🏻

gziolo commented on packages/create-block-tutorial-template/plugin-templates/$slug.php-dynamic.mustache:
```diff
@@ -0,0 +1,60 @@
+<?php
```
We probably don't need this file anymore.

ryanwelcher replied:
nice catch! I updated all of the files in the tutorial template to use the conditionals

gziolo reviewed (commented):
This looks great overall. I'm really excited about this new flag that gives more flexibility for block developers.

Should we also include a prompt from CLI when using the interactive mode? I'm not sure at the moment. This option won't have any impact on 3rd party templates for start...

ryanwelcher committed:
Updating the create-block-tutorial-template to use the mustache conditionals.

ryanwelcher committed:
Add the --webpack-copy-php to the scripts based on the the --is-dynamic flag.

ryanwelcher commented:
> Should we also include a prompt from CLI when using the interactive mode? I'm not sure at the moment. This option won't have any impact on 3rd party templates for start...

I'm not sure here either, my first thought is "not yet". As you mentioned, this is really only going to work with templates that adopt the approach we've taken in the internal templates. Maybe once there is a way to determine if the template supports it we can add the prompt?

ryanwelcher committed:
Move the .trim call into the context of the if statement. Proof that @gziolo is a much better developer than I am.

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

gziolo commented on packages/create-block/CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 
 ## Unreleased
 
+-	Introduce `--is-dynamic` flag to allow creation of dynamic blocks if the template supports it. ([#41289](https://github.com/WordPress/gutenberg/pull/41289)).
```
```suggestion
### New Feature

-   Introduce `--is-dynamic` flag to allow creation of dynamic blocks if the template supports it ([#41289](https://github.com/WordPress/gutenberg/pull/41289)).
```

gziolo committed:
Update packages/create-block/CHANGELOG.md

gziolo commented on packages/create-block/lib/index.js:
```diff
@@ -106,6 +109,7 @@ program
 					const blockPrompts = getPrompts( pluginTemplate, [
 						'slug',
 						'namespace',
+
```
```suggestion
```

gziolo committed:
Update packages/create-block/lib/index.js

gziolo commented on packages/create-block-tutorial-template/plugin-templates/$slug.php.mustache:
```diff
@@ -37,7 +37,35 @@
  *
  * @see https://developer.wordpress.org/reference/functions/register_block_type/
  */
+{{^isDynamic}}
```
Not a blocker. Should we inject also `isStatic` variable to make those checks easier to follow? Example:
```suggestion
{{#isStatic}}
```

gziolo commented:
One minor thing I discovered while testing is that `src/save.js` files get included when using `--is-dynamic` flag. They aren't referenced in the code so we can safely remove them by wrapping them with the condition like in the `es5` template. Files that should be updated:

https://github.com/WordPress/gutenberg/blob/304eee65580ff8aeb0ceb586c76a60d5c7d4f5a4/packages/create-block/lib/templates/block/save.js.mustache
https://github.com/WordPress/gutenberg/blob/304eee65580ff8aeb0ceb586c76a60d5c7d4f5a4/packages/create-block-tutorial-template/block-templates/save.js.mustache

In the case of the tutorial template, we also should apply two other updates:


`save` and its import should be behind the flag:
https://github.com/WordPress/gutenberg/blob/cffede0a17bad9bb3cf5b3da1a8070fcc54d4e87/packages/create-block-tutorial-template/block-templates/index.js.mustache#L44-L47

The other tricky bit is the `attributes`. In the case of the dynamic block in the tutorial template, we should have a bit different version than:

https://github.com/WordPress/gutenberg/blob/cffede0a17bad9bb3cf5b3da1a8070fcc54d4e87/packages/create-block-tutorial-template/index.js#L14-L20

The output on the frontend also doesn't care about the value provided in the input field:

<img width="650" alt="Screenshot 2022-06-15 at 09 18 07" src="https://user-images.githubusercontent.com/699132/173766833-773ae3eb-7a85-425c-b481-8f178cfa1642.png">

In general, dynamic blocks in almost every case would source the attribute from the metadata in the HTML comment rather than from the HTML itself.

Maybe, we should skip adding a dynamic version option for the tutorial?

gziolo commented:
The other thing that I was thinking about is how we introduce the new CLI option:

> 4. Generating a dynamic block based on the built-in template.
>
> ```bash
> $ npx @wordpress/create-block --is-dynamic
> ```

In this form, we sort of imply that all templates are static by default. It doesn't have to be true. 

> > Should we also include a prompt from CLI when using the interactive mode? I'm not sure at the moment. This option won't have any impact on 3rd party templates for start...
> 
> I'm not sure here either, my first thought is "not yet". As you mentioned, this is really only going to work with templates that adopt the approach we've taken in the internal templates. Maybe once there is a way to determine if the template supports it we can add the prompt?

What if we would approach it differently and name it `variant` (or something along those lines):

```bash
$ npx @wordpress/create-block --variant=dynamic
$ npx @wordpress/create-block --variant=static
```

Then we could have in the template definition optional variants:

```js
module.exports = {
	defaultValues: {
		variant: 'static',
	},
	variants: [ 'static', 'dynamic', 'whatever' ],
};
```

In the template we could inject variables for every variant:
- `isStaticVariant`
- `isDynamicVariant`
- `isWhateverVariant`

We could also use this array, when present, to generate the prompt for selecting variants from CLI when using an interactive mode.

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge branch 'trunk' of github.com:ryanwelcher/gutenberg into trunk

ryanwelcher committed:
Remove extra code block added in error.

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

gziolo commented:
I did quick prototyping, and it looks like adding `renderTemplate` to `block.json` is straightforward even in Gutenberg: #42430.

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge branch 'trunk' into try/adding-is-dynamic-flag

ryanwelcher committed:
Merge branch 'try/adding-is-dynamic-flag' of github.com:ryanwelcher/gutenberg into try/adding-is-dynamic-flag

ryanwelcher commented:
> What if we would approach it differently and name it `variant` (or something along those lines):
> 
> ```shell
> $ npx @wordpress/create-block --variant=dynamic
> $ npx @wordpress/create-block --variant=static
> ```
> 
> Then we could have in the template definition optional variants:
> 
> ```js
> module.exports = {
> 	defaultValues: {
> 		variant: 'static',
> 	},
> 	variants: [ 'static', 'dynamic', 'whatever' ],
> };
> ```
> 
> In the template we could inject variables for every variant:
> 
> * `isStaticVariant`
> * `isDynamicVariant`
> * `isWhateverVariant`
> 
> We could also use this array, when present, to generate the prompt for selecting variants from CLI when using an interactive mode.

This feels like the correct approach. I don't mind `--variant` as the name. What do you think of  `--block-type` or just `--type`? We tend to refer to static and dynamic bocks as types and it that sense it's closer to the mental model we're putting out. I don't feel too strongly either way,

ryanwelcher commented:
I've been thinking a little more about the name and I think that`--variant` may be more flexible. 
Besides`static` and `dynamic`, there is any number of variations that may not be tied to the block type - for example, `es5`, and `typescript` either of these styles of variants could be static or dynamic but allow for further customization of templates.

ryanwelcher committed:
Adding variants array to the template and generating template variables for each one.

ryanwelcher committed:
Add isStatic check to save.js

ryanwelcher committed:
Add logic to return empty object if no variants are defined in the template

ryanwelcher committed:
Update the template variable names to be is{variantName}Variant.

ryanwelcher committed:
Update templates to use new variable name convention and look for the variable name for each type rather than checking for the negation.

ryanwelcher committed:
If no variant is select via CLI, use the first one in the array of defined variants as the default.

ryanwelcher committed:
Check that the passed variant is a valid variant and exit if not.

ryanwelcher committed:
Define a single variant for the create-block-tutorial-template. Remove references to the dynamic block version.

ryanwelcher committed:
Add template selection to interactive mode.

ryanwelcher commented:
I've converted the flag to `--variant` and introduced a `variants` property to the templates. Anything added to the array will generate template variables that can be used inside the templates to control what is rendered. For example

```js
variants: [ 'static', 'dynamic']
```
Would create the variables `isStaticVariant` and `isDynamicVariant` that would be available to the mustache templates.

If there is no variant passed, we use the first one in the list as the default and if there is an invalid variant passed, the script will error out with a message

![invalid_variant](https://user-images.githubusercontent.com/1259027/184949311-0f546c5f-e767-4991-89e3-8197f7718ef0.png)

I've also added a prompt during interactive mode that retrieves the list of variants:

![variants](https://user-images.githubusercontent.com/1259027/184950422-1e4a4234-b33d-4f2a-9b52-8f0c5a6232d9.png)

It should be noted all of this depends on the mustache templates using the provided variables. The template can define all of the variants it wants but if the `is{Type}Variation` vars are not used to conditional render output, the template will still render. This means that any existing templates should continue to work as expected

ryanwelcher commented:
As for the `@wordpress/create-block-template-project`, I've defined a single variation for it, `static`,  and removed the reference to the dynamic variation we tried to add because we can't easily change the attribute definitions as @gziolo mentioned in his [comment above](https://github.com/WordPress/gutenberg/pull/41289#issuecomment-1156082662)

ryanwelcher committed:
Update changelog.

gziolo commented on packages/create-block/lib/templates/plugin/$slug.php.mustache:
```diff
@@ -37,7 +37,35 @@
+{{/isStaticVariant}}
+{{#isDynamic}}
+function {{namespaceSnakeCase}}_{{slugSnakeCase}}_block_init() {
+	register_block_type(
```
We could branch for static vs dynamic only for the block registration here.

ryanwelcher replied:
Yes, I found that it ends up creating files with odd whitespace when I tried to get tricky with it.

gziolo commented on packages/create-block/lib/templates/es5/$slug.php.mustache:
```diff
@@ -68,7 +68,7 @@ function {{namespaceSnakeCase}}_{{slugSnakeCase}}_block_init() {
 		filemtime( "$dir/$style_css" )
 	);
-
+	{{^isDynamic}}
```
Nit:
```suggestion
	{{#isStaticVariant}}
```

gziolo commented on packages/create-block/lib/templates/plugin/$slug.php.mustache:
```diff
@@ -37,7 +37,35 @@
+	require plugin_dir_path( __FILE__ ) . 'build/template.php';
+	return ob_get_clean();
+}
+{{/isDynamic}}
```
Here and in other places I see old usage of `isDynamic`, it should be `isDynamicVariant` with the latest changes applied:
```suggestion
{{/isDynamicVariant}}
```

gziolo commented on packages/create-block/lib/templates/es5/$slug.php.mustache:
```diff
@@ -77,5 +77,33 @@ function {{namespaceSnakeCase}}_{{slugSnakeCase}}_block_init() {
 		)
 	);
+	{{/isDynamic}}
+	{{#isDynamic}}
```
Here and in other places in this file:

```suggestion
	{{#isDynamicVariant}}
```

gziolo commented on packages/create-block/lib/templates.js:
```diff
@@ -222,15 +225,43 @@ const getDefaultValues = ( pluginTemplate ) => {
 };
 
+const getTemplateVariantVars = ( variants, variant ) => {
+	const variantTypes = {};
```
I wasn't sure what variant types mean at first, maybe we could rename the variable to:
```suggestion
	const variantVars = {};
```

This would align with the function name, what do you think?

gziolo commented on packages/create-block/lib/init-package-json.js:
```diff
@@ -24,6 +24,7 @@ module.exports = async ( {
 	npmDependencies,
 	npmDevDependencies,
 	customScripts,
+	isDynamic,
```
Here and in other places in the file:
```suggestion
	isDynamicVariant,
```

gziolo commented on packages/create-block/lib/index.js:
```diff
@@ -111,6 +114,7 @@ program
 						! Object.keys( optionsValues ).includes( name );
 					const blockPrompts = getPrompts( pluginTemplate, [
 						'slug',
+						'variant',
```
I think we should include the variant prompt only when `pluginTemplate.variants` contains more than one option. If there is a single option, we can automatically set it. Unless `inquirer` automatically skips the prompt in case there are less than 2 options.

We can test it with other templates that don't have variants defined.

ryanwelcher replied:
`inquirer` showed the prompt even with only a single option. I have added some logic to only promptly if there are more than one variant defined.

gziolo commented on packages/create-block/lib/templates.js:
```diff
@@ -44,6 +44,7 @@ const predefinedPluginTemplates = {
 				html: false,
 			},
 		},
+		variants: [ 'static', 'dynamic' ],
```
We should also add variants to the `es5` template definition.

The other challenge is that both templates provided were static so far. We will have to rename them to something more general. I don't think we want to keep the static vs dynamic distinction in the block definition anymore.

ryanwelcher replied:
I've added the variations to the es5 template and removed "static" from the various properties. I have also renamed the `static` template to `standard`. Not 100% on the name though, @KevinBatdorf suggested `default` in my stream which I don't mind either. Any thoughts?

gziolo replied:
`default` sounds good 👍🏻

gziolo reviewed (commented):
Thank you for keeping the work on this feature. I left my feedback. It's shaping up nicely. We will have to test it with some existing templates that don't support variants to ensure people don't see an empty prompt for variants.

One thing that we could also consider is to set the `dynamic` variant as the default for the block. This way, we could bring back i18n support to the `save` function.

gziolo commented on packages/create-block/README.md:
```diff
@@ -50,6 +50,7 @@ Options:
 --no-wp-scripts              disable integration with `@wordpress/scripts` package
 --wp-env                     enable integration with `@wordpress/env` package
 -h, --help                   output usage information
+--is-dynamic                 generates a dynamic block based on the template being used. The template must opt-in to this functionality.
```
It's `--variant` now, and the description needs to be rephrased.

gziolo commented on packages/create-block/README.md:
````diff
@@ -72,7 +73,13 @@ $ npx @wordpress/create-block --template my-template-package
+4. Generating a dynamic block based on the built-in template.
+
+```bash
+$ npx @wordpress/create-block --is-dynamic
````
```suggestion
$ npx @wordpress/create-block --variant dynamic
```

ryanwelcher committed:
Update isDynamic to isDynamicVariant

ryanwelcher committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

ryanwelcher committed:
Merge branch 'trunk' into try/adding-is-dynamic-flag

ryanwelcher committed:
Replace variantTypes with variantVars

ryanwelcher committed:
Update isDynamic to isDynamicVariant... again.

ryanwelcher committed:
Update readme to use --variant flag.

ryanwelcher committed:
Add logic to conditionally show the variant prompt only if there are more than one.

ryanwelcher committed:
Adds variants to the es5 block and removes "static" from various properties.

ryanwelcher committed:
Change "static" template to "standard".

ryanwelcher committed:
Convert variants to an object where the property is the variant name and it's value is settings that will override the defaults.

ryanwelcher committed:
Define a dynamic variant for the create-block-tutorial and override the attribute settings.

ryanwelcher commented:
@gziolo I've addressed your notes from the code review. I've also made a change to how we use `variants`. They are now stored in an object. This allows us to define variant-specific values which override the defaults and addresses the issue we were having with the create-block-project-template.

gziolo commented on packages/create-block/lib/scaffold.js:
```diff
@@ -99,6 +100,8 @@ module.exports = async (
 		editorStyle,
 		style,
+		...getTemplateVariantVars( variants, variant ),
+		...variants[ variant ],
```
That setup will probably override the input provided by the user. It might be a little bit tricky to propagate the default values for the variant selected correctly. Actually, if you would force the user to pick the variant for the block first, then you could tweak the default values passed to prompts and that would cover everything.

gziolo replied:
I was right. See the screenshot:

<img width="1181" alt="Screenshot 2022-08-19 at 14 19 41" src="https://user-images.githubusercontent.com/699132/185636599-b0bc45a8-9c25-404f-909c-d670c2b2e165.png">

I selected the default value hinted in the prompt. It wasn't the one coming from the variant. However, at the end of the scaffolding, the values from the variant were used instead and that's why the folder name is different (on the right side in the explored).

ryanwelcher replied:
@gziolo I believe I have addressed this. If you could test it again when have have a chance. Thanks!

gziolo commented on packages/create-block/lib/templates.js:
```diff
@@ -21,29 +21,48 @@ const prompts = require( './prompts' );
 		},
 		templatesPath: join( __dirname, 'templates', 'es5' ),
+		variants: {
+			static: {
+				slug: 'example-static-es5',
+				title: 'Example Static Block (ES5)',
+			},
```
The values for the `static` variant could be set directly in the `defaultValues` object.

The same applies to the templates related to the `standard` template.

gziolo replied:
I was pondering one more optimization from the perspective of the developer writing a scaffolding template. What if we assume that everything that is included in `defaultValues` is the default variant, so we never really need to pick it explicitly? If there are any variants, then we use the `variants` option, but we only provide non-default variants:

```js
variants: {
 	dynamic: {
 		slug: 'example-dynamic-block',
 		title: 'Example Dynamic Block',
 	},
 	anotherVariant: {},
},
```

That would generate 3 vars to use with the template:
- `isDefaultVariant`
- `isDynamicVariant`
- `isAnotherVariant`

If no variants were included, we wouldn't have any variables injected into templates.

This would also remove some checks like whether there is more than one variant provided because we would always have the default one and one or more in the `variants` field.

The only drawback is that we wouldn't have a way to pass the alternate name for the default variant so users won't see the radio option like:

- [x] static
- [ ] dynamic
- [ ] anotherVariant

Instead, it would look like

- [x] default
- [ ] dynamic
- [ ] anotherVariant

ryanwelcher replied:
I like the idea, perhaps we can iterate on it?

gziolo replied:
I moved the default values from `static` to `defaultValues` in https://github.com/WordPress/gutenberg/pull/43481. I didn't refactor further. It's still valuable to have a way to name the default variant. In the future, we could also figure out how to pass some sort of description for the variant. At the moment it's `static` vs `dynamic` which might not be clear for someone starting with WP blocks.

gziolo commented:
>  I've also made a change to how we use variants. They are now stored in an object. This allows us to define variant-specific values which override the defaults and addresses the issue we were having with the create-block-project-template.

Excellent follow-up @ryanwelcher. Brilliant idea to use an object with overrides for variants 💯 

We now only need to settle on the final shape of variants, and we should be good to go. I still need to re-test the plugin after applying all the changes.

gziolo commented:
Code-wise everything looks good. There is one issue that I flagged in https://github.com/WordPress/gutenberg/pull/41289/files#r949815229 that I consider a blocker so it needs to get resolved.

ryanwelcher committed:
Get the variant first and then retrieve the defaults based on that value.

ryanwelcher commented:
@gziolo 

> Code-wise everything looks good. There is one issue that I flagged in https://github.com/WordPress/gutenberg/pull/41289/files#r949815229 that I consider a blocker so it needs to get resolved.

I believe I have addressed this. Please retest when you can.

gziolo committed:
Update templates.js

gziolo committed:
Update templates.js

gziolo reviewed (approved):
It looks good. Let's get this one in after all those iterations 🎉

I'm running some extensive testing and open follow-up PR with enhancements if necessary.

gziolo commented on packages/create-block/lib/prompts.js:
```diff
@@ -134,8 +134,16 @@ const updateURI = {
+const variant = {
+	type: 'list',
+	name: 'variant',
+	message: 'The template variant to use for this block.',
```
```suggestion
	message: 'The template variant to use for this block:',
```

gziolo committed:
Update packages/create-block/lib/prompts.js

gziolo commented on packages/create-block/lib/index.js:
```diff
@@ -107,14 +113,30 @@ program
-					] ).filter( filterOptionsProvided );
+
+					// Get the variant prompt first. This will help get the default values
+					const variantPrompt =
```
When I provide the variant from the CLI, it asks again for the variant. We can address it seperately.

gziolo replied:
Fixed in https://github.com/WordPress/gutenberg/pull/43481.
